pilot/pkg/networking/core/v1alpha3/listener_inbound.go

```diff
@@ -0,0 +1,674 @@
+
+		if cc.bindToPort {
+			// If this config is for bindToPort, we want to actually create a real Listener.
+			listeners = append(listeners, lb.inboundCustomListener(cc, chains))
+		} else {
+			// Otherwise, just append the filter chain to the virtual inbound chains.
+			virtualInboundFilterChains = append(virtualInboundFilterChains, chains...)
+		}
+	}
+
+	if lb.node.GetInterceptionMode() != model.InterceptionNone {
+		// Prepend virtual inbound, as long as we are using redirection.
+		listeners = append([]*listener.Listener{lb.inboundVirtualListener(virtualInboundFilterChains)}, listeners...)
+	}
+
+	return listeners
+}
+
+// inboundVirtualListener builds the virtual inbound listener.
+func (lb *ListenerBuilder) inboundVirtualListener(chains []*listener.FilterChain) *listener.Listener {
+	actualWildcard, _ := getActualWildcardAndLocalHost(lb.node)
+
+	// Build the "virtual" inbound listener. This will capture all inbound redirected traffic and contains:
+	// * Passthrough filter chains, matching all unmatched traffic. There are a few of these to handle all cases
+	// * Service filter chains. These will either be for each Port exposed by a Service OR Sidecar.Ingress configuration.
+	vi := &listener.Listener{
+		Name:             model.VirtualInboundListenerName,
+		Address:          util.BuildAddress(actualWildcard, ProxyInboundListenPort),
+		TrafficDirection: core.TrafficDirection_INBOUND,
+	}
+	if lb.node.GetInterceptionMode() == model.InterceptionTproxy {
+		vi.Transparent = proto.BoolTrue
+	}
+	if lb.node.Metadata.InboundListenerExactBalance {
+		vi.ConnectionBalanceConfig = &listener.Listener_ConnectionBalanceConfig{
+			BalanceType: &listener.Listener_ConnectionBalanceConfig_ExactBalance_{
+				ExactBalance: &listener.Listener_ConnectionBalanceConfig_ExactBalance{},
+			},
+		}
+	}
+	accessLogBuilder.setListenerAccessLog(lb.push, lb.node, vi, istionetworking.ListenerClassSidecarInbound)
+
+	// First, we setup the catchall filter chains. These are present in all proxies and catch any
+	// traffic that is received on a port not exposed in the service, passing it through.
+	vi.FilterChains = append(vi.FilterChains, buildInboundPassthroughChains(lb)...)
+	vi.FilterChains = append(vi.FilterChains, chains...)
+
+	// Listener filters are derived from the filter chains, so we set them up after filter chain construction
+	vi.ListenerFilters = populateListenerFilters(lb.node, vi, false)
+	vi.ListenerFiltersTimeout = getProtocolDetectionTimeout(lb.push.Mesh)
+	vi.ContinueOnListenerFiltersTimeout = true
+
+	return vi
+}
+
+func (lb *ListenerBuilder) inboundCustomListener(cc inboundChainConfig, chains []*listener.FilterChain) *listener.Listener {
+	l := &listener.Listener{
+		Name:             cc.Name(istionetworking.ListenerProtocolTCP),
+		Address:          util.BuildAddress(cc.bind, cc.port.TargetPort),
+		TrafficDirection: core.TrafficDirection_INBOUND,
+	}
+	if lb.node.Metadata.InboundListenerExactBalance {
```

Comment:
may be we should extract this in to a function like `lb.maybeApplyInboundExactBalance()` and use it every where?